### DISCUSSION TITLE
highlight a line using ggproto.

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#12] [Q&A]
@yjunechoe alerted me to this question... https://rstats.me/@danwwilson/112373592162157242 with this first stab.  https://gist.github.com/danwwilson/cb2814fcf275c28cf5c9d988fd3c5713

But a compute_panel computation approach seems to work...  https://evamaerey.github.io/mytidytuesday/2024-05-06-gghighlightline/gghighlightline.html but I'd be interested re-visiting with a compute_group approach to follow more closely with this 'However, as a general rule it is better to begin by modifying compute_group() only and see if the performance is adequate.' from ggplot2 book, extensions chapter.

**Comment by teunbrand** [comment]
Have you taken a look at how {gghighlight} does this?

**Reply by yjunechoe** [reply]
BTW following up on this, I re-discovered one of Yutani's blog post which goes into why "just highlight X" is one of those "some easy things are hard" cases in ggplot/ggproto - https://yutani.rbind.io/post/2017-11-07-ggplot-add/. @EvaMaeRey maybe this is of interest given our discussion on mastodon

@teunbrand I'm curious if you have any thoughts/updates on the assessment of this problem (as best solved by `ggplot_add()`) from the perspective of modern ggplot, 7 years after this was written?

**Reply by EvaMaeRey** [reply]
@yjunechoe thanks for passing along! for my shallow quick look, below, I was impressed gghighlight seems to display a lot of mid plot build self-awareness and self-editing that's orthogonal to other gg behavior, which seemed like it would be hard to pull off.

**Reply by yjunechoe** [reply]
Yeah! [Apparently](https://twitter.com/yutannihilat_en/status/1522412049061474304?t=L0OYiEGQM8X3uhzd6fj5ZA&s=19) all the logic lives inside [ggplot_add.gg_highlighter()](https://github.com/yutannihilation/gghighlight/blob/1c4931f32521f3f25c0f5cc00595baec241b8280/R/gghighlight.R#L129-L279) which is simultaneously fascinating and terrifying to look at 😅

**Reply by EvaMaeRey** [reply]
'fascinated and terrified' and 'waaa?' were definitely my superficial feelings!  

I did write this revised Stat approach with grouping and highlight_condition aes, which I think might feel friendlier, given feedback from Dan and the gghighlight experience:
 https://evamaerey.github.io/mytidytuesday/2024-05-06-gghighlightline/gghighlightline2.html 

It feels a little funny not to be able to just use group as aes, but I couldn't manage it so went for 'grouping'.

**Reply by teunbrand** [reply]
These days, I'd probably go for a <Layer> subclass approach so that I can take control over `Layer$setup_layer()`.
I have definitely abused `ggplot_add()` as well, like [here](https://github.com/teunbrand/ggh4x/blob/80c649109d2ce9ef1bb690f31f366f4a259a4d3a/R/facetted_pos_scales.R#L186), [here](https://github.com/teunbrand/ggh4x/blob/80c649109d2ce9ef1bb690f31f366f4a259a4d3a/R/scale_listed.R#L142), and [here](https://github.com/teunbrand/ggh4x/blob/80c649109d2ce9ef1bb690f31f366f4a259a4d3a/R/force_panelsize.R#L94), but I don't recommend it unless absolutely necessary.

Really barebones example of layer subclassing to just subset the data using an expression.

``` r
library(ggplot2)

layer_subset <- function(..., subset) {
  parent <- layer(...)
  ggproto(
    NULL, parent,
    subset = enexpr(subset),
    setup_layer = function(self, data, plot) {
      i <- rlang::eval_tidy(self$subset, data = data)
      data <- vctrs::vec_slice(data, i)
      ggproto_parent(parent, self)$setup_layer(data, plot)
    }
  )
}

ggplot(mpg, aes(displ, hwy)) +
  layer_subset(
    geom = GeomPoint,
    stat = StatIdentity,
    position = PositionIdentity,
    subset = displ > 5
  )
```

![](https://i.imgur.com/DzvseQO.png)

<sup>Created on 2024-05-14 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by yjunechoe** [reply]
That's really clever! Never even crossed my mind to extend Layer, but `$setup_data()` makes lotta sense for this

BTW is the `parent <- layer(...)` trick used bc `Layer` isn't exported? (and on that point, is `Layer` formally part of the extension API? I feel like there was a discussion of this few years ago...)

**Reply by teunbrand** [reply]
> is the parent <- layer(...) trick used bc Layer isn't exported?

Yeah that is right. I don't think `Layer` is officially open for extensions as it isn't advertised anywhere, but neither is `ggplot_add()`. I'm unsure why <Layer> isn't exported.

**Comment by EvaMaeRey** [comment]
I'm not sure exactly how it does it's thing (split layer into two).  It does seem very usable suited for the case in question.  It seems like the documentation could benefit from a few examples for the very basic condition (the group which is "New Zealand") in addition to the value based conditions.


``` r
library(tidyverse)

gapminder::gapminder %>% 
  filter(continent == "Oceania") %>% 
  filter(year > 1995) %>% 
  ggplot() + 
  aes(x = year, y = lifeExp, group = country, color = country) + 
  geom_line()
```

![](https://i.imgur.com/tAMfD3R.png)

``` r

last_plot()$layers |> length()
#> [1] 1

last_plot() +
  gghighlight::gghighlight(country %in% c("New Zealand"), 
                           unhighlighted_colour = "darkgrey", 
                           use_direct_label = F) 
#> Warning: The `unhighlighted_colour` argument of `gghighlight()` is deprecated as of
#> gghighlight 0.2.0.
#> ℹ Please use the `unhighlighted_params` argument instead.
#> This warning is displayed once every 8 hours.
#> Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
#> generated.
#> Warning: Tried to calculate with group_by(), but the calculation failed.
#> Falling back to ungrouped filter operation...
```

![](https://i.imgur.com/YrvvP6G.png)

``` r

last_plot()$layers
#> [[1]]
#> mapping: x = ~highlight..........1, y = ~highlight..........2, colour = ~highlight..........3, group = ~highlight..........group, fill = NULL 
#> geom_line: na.rm = FALSE, orientation = NA
#> stat_identity: na.rm = FALSE
#> position_identity 
#> 
#> [[2]]
#> mapping: x = ~year, y = ~lifeExp, colour = ~country, group = ~country 
#> geom_line: na.rm = FALSE, orientation = NA
#> stat_identity: na.rm = FALSE
#> position_identity

layer_data(i = 1) 
#>      x      y group PANEL flipped_aes   colour linewidth linetype alpha
#> 1 1997 78.830     1     1       FALSE darkgrey       0.5        1    NA
#> 2 2002 80.370     1     1       FALSE darkgrey       0.5        1    NA
#> 3 2007 81.235     1     1       FALSE darkgrey       0.5        1    NA
#> 4 1997 77.550     2     1       FALSE darkgrey       0.5        1    NA
#> 5 2002 79.110     2     1       FALSE darkgrey       0.5        1    NA
#> 6 2007 80.204     2     1       FALSE darkgrey       0.5        1    NA

layer_data(i = 2)
#>    colour    x      y group PANEL flipped_aes linewidth linetype alpha
#> 1 #F8766D 1997 77.550     1     1       FALSE       0.5        1    NA
#> 2 #F8766D 2002 79.110     1     1       FALSE       0.5        1    NA
#> 3 #F8766D 2007 80.204     1     1       FALSE       0.5        1    NA

gapminder::gapminder %>% 
  filter(continent == "Americas") %>% 
  ggplot() + 
  aes(x = year, y = lifeExp, group = country, color = country) +
  geom_line() +
  scale_color_viridis_d() +
  gghighlight::gghighlight(country %in% c("Bolivia", "Chile"), 
                           unhighlighted_colour = "darkgrey", 
                           use_direct_label = F) 
#> Warning: Tried to calculate with group_by(), but the calculation failed.
#> Falling back to ungrouped filter operation...
```

![](https://i.imgur.com/PZ7qJch.png)

``` r

last_plot() +
  scale_color_manual(values = c("orange", "orange")) + 
  facet_wrap(~country)
#> Scale for colour is already present.
#> Adding another scale for colour, which will replace the existing scale.
```

![](https://i.imgur.com/n3GeWHN.png)

<sup>Created on 2024-05-06 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
The gghighlight approach to faceting by group with all-groups shadow feels more intuitive than user two-layer approach (drop faceting variable in first layer) which I learned somewhere.  

``` r
library(tidyverse)
palmerpenguins::penguins |>
  ggplot(aes(x = bill_length_mm, y = bill_depth_mm)) +
  facet_wrap(~ species) +
  ggstamp::theme_void_fill("linen") +
# plot all data in each species facet
  geom_point(data =
               . %>% select(-species),                           # drop faceting var in first layer
             size = 3, shape = 21, stroke = .7,
             color = "burlywood4", fill= "white",
             alpha = .7) +
# use global data and plot each species in its facet
  geom_point(size = 3, shape = 21, stroke = .2,
             color = "white", fill = "burlywood4")
#> Warning: Removed 6 rows containing missing values or values outside the scale range
#> (`geom_point()`).
#> Warning: Removed 2 rows containing missing values or values outside the scale range
#> (`geom_point()`).
```

![](https://i.imgur.com/xygf6wJ.png)

``` r

layer_data(i = 1) %>% head()
#>      x    y PANEL group shape     colour size  fill alpha stroke
#> 1 39.1 18.7     1    -1    21 burlywood4    3 white   0.7    0.7
#> 2 39.5 17.4     1    -1    21 burlywood4    3 white   0.7    0.7
#> 3 40.3 18.0     1    -1    21 burlywood4    3 white   0.7    0.7
#> 4   NA   NA     1    -1    21 burlywood4    3 white   0.7    0.7
#> 5 36.7 19.3     1    -1    21 burlywood4    3 white   0.7    0.7
#> 6 39.3 20.6     1    -1    21 burlywood4    3 white   0.7    0.7
```

<sup>Created on 2024-05-15 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by EvaMaeRey** [comment]
Defining compute_panel here and elsewhere seems attractive because it seems to match how I and I expect other think about it.  I guess I'm wondering if anyone can expand upon this?  'However, as a general rule it is better to begin by modifying compute_group() only and see if the performance is adequate.'

**Reply by yjunechoe** [reply]
I talked about that issue a bit in one of my blog posts - https://yjunechoe.github.io/posts/2022-07-06-ggplot2-delayed-aes-2/#other-compute_-extensions

A good example of what the quote is referring to (and one I explicitly talk about in the blog post) is something like the `compute_*()` method in `StatIdentity`. The behavior of the stat is to return the data as-is, and that can happen at any `compute_*()` function. But if you're not doing anything with the data, you might as well return early from `$compute_layer()` immediately when you receive it, and not have it be split all the way down to `$compute_group()` only for nothing to happen and then combine the data later. Indeed, that's why `StatIdentity` extends `$compute_layer()` and not `$compute_group()` for reasons of performance, parsimony, and theoretical purity. But in practice I think those are rather small matters since Stats don't do much heavy computation to begin with, hence the qualifier "see if the performance is *adequate*"

I like to think of this in terms of `{dplyr}` - imagine if you want a function that "computes" on a data frame column to add 1 to it. You could write it like this:

```r
library(dplyr)
df <- tibble(group = gl(2, 2), value = sample(4))
df

f1 <- function(x) {
  x %>% 
    mutate(value = value + 1)
}
```

Or like this:

```r
f2 <- function(x) {
  x %>% 
    group_by() %>% 
    mutate(value = value + 1) %>% 
    ungroup()
}
```

And the result is `identical(f1(df), f2(df))`, but you'd prefer `f1()` because it's simpler. `f2()` does a split-apply-combine that's redundant for the kind of computation you're doing (versus, for example, `scale()`-ing by group).

---

I think that a tl;dr to your question, if I understand correctly, is that "start from `$compute_group()`" is a good heuristic because you can never go wrong with implementing the narrowest form of the solution. But the quote encourages you to streamline the solution afterwards, in case it's more appropriate to implement the logic at a higher level of abstraction (like panel or layer)

**Reply by EvaMaeRey** [reply]
Thanks for directing me to your blog post (again!).  I think it's probably worth printing out at this point!
